src/server/pkg/storage/fileset/merge.go

```diff
@@ -28,153 +27,169 @@ func newMergeReader(chunks *chunk.Storage, fileSets []FileSet) *MergeReader {
 }
 
 // Iterate iterates over the files in the merge reader.
-func (mr *MergeReader) Iterate(ctx context.Context, cb func(File) error) error {
+func (mr *MergeReader) Iterate(ctx context.Context, cb func(File) error, deletive ...bool) error {
+	if len(deletive) > 0 && deletive[0] {
+		return mr.iterateDeletive(ctx, cb)
+	}
+	return mr.iterate(ctx, cb)
+}
+
+func (mr *MergeReader) iterate(ctx context.Context, cb func(File) error) error {
 	var ss []stream
 	for _, fs := range mr.fileSets {
+		ss = append(ss, &fileStream{
+			iterator: NewIterator(ctx, fs, true),
+			priority: len(ss),
+			deletive: true,
+		})
 		ss = append(ss, &fileStream{
 			iterator: NewIterator(ctx, fs),
 			priority: len(ss),
 		})
 	}
 	pq := newPriorityQueue(ss)
+	var skipPrefix string
+	var skipPriority int
 	return pq.iterate(func(ss []stream, _ ...string) error {
-		// Collect files from file streams.
-		var files []File
+		var fss []*fileStream
 		for _, s := range ss {
-			fs := s.(*fileStream)
-			files = append(files, fs.file)
+			fss = append(fss, s.(*fileStream))
+		}
+		if skipPrefix != "" && strings.HasPrefix(fss[0].key(), skipPrefix) {
+			for i, fs := range fss {
+				if fs.streamPriority() < skipPriority {
+					fss = fss[:i]
+					break
+				}
+			}
 		}
-		if len(files) == 1 {
-			return cb(files[0])
+		if len(fss) == 0 {
+			return nil
+		}
+		if IsDir(fss[0].key()) {
```

Comment:
I didn't think there were directories at this point?

Reply:
Yeah, but directory deletions are still captured in the deletive set as directory paths. Deleting a directory means deleting all of the files underneath it.